File: Daily/PY/Leetcode150-逆波兰表达式.py
from typing import List
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stk = []
        mat = ["+", "-", "*", "/"]

        def eval(s):
            b = stk.pop()
            a = stk.pop()
            if s == "+":
                stk.append(a + b)
            elif s == "-":
                stk.append(a - b)
            elif s == "*":
                stk.append(a * b)
            else:
                stk.append(int(a / b))

        for tt in tokens:
            if tt in mat:
                eval(tt)
            else:
                stk.append(int(tt))

        return stk.pop()

File: Daily/sbgo/Leetcode150-逆波兰表达式.go
package sbgo

import "strconv"

func evalRPN(tokens []string) int {
	stk := []int{}
	eval := func(s string) {
		a, b := stk[len(stk)-2], stk[len(stk)-1]
		stk = stk[:len(stk)-2]
		if s == "+" {
			stk = append(stk, a+b)
		} else if s == "-" {
			stk = append(stk, a-b)
		} else if s == "*" {
			stk = append(stk, a*b)
		} else {
			stk = append(stk, a/b)
		}
	}
	for _, tt := range tokens {
		if tt == "+" || tt == "-" || tt == "*" || tt == "/" {
			eval(tt)
		} else {
			num, err := strconv.Atoi(tt)
			if err == nil {
				stk = append(stk, num)
			}
		}
	}
	return stk[0]
}
